### PR TITLE
Project 2

### PR DESCRIPTION
**Pull request by SusanaPerezJ**

### CODE REVIEW

File: portfolio/src/main/webapp/index.html [hunk 1]
```diff
@@ -28,12 +28,16 @@ <h2>Pictures!</h2>
       <p>I was there for a just a couple of days but I had a lot of fun, this is me visiting Marina del Rey's pier.</p>
       <img src="images/beach.jpg" alt="Sunset at sea" class="standard-size image-on-center"/>
       <p>I took this picture on my last day in California, I really enjoyed seeing the waves and the orange color of the sky.</p>
-      <h2>Random Facts about me </h2> 
+      <h2>Random Facts about me</h2> 
       <p>Click this button to see a random fact about me:</p>
       <button onclick="addRandomFact()">Click Here!</button>
       <div id="fact-container"></div>
+      <br><br/>
```
Comment (wait321) [inline portfolio/src/main/webapp/index.html]: Did you mean `<br></br>`? In any case br tags just add a line break immediately, so we always do `<br />`, which is the same as `<br></br>`, to save space.

**Review by wait321 — CHANGES_REQUESTED** [review]
One minor issue with HTML, but otherwise looks good.

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -26,3 +26,9 @@
      factContainer.innerText = randomFact;
  }
 
+//Fetches a message from the server and adds it to the DOM
+function getFetchedMessage(){
+    fetch('/data').then(response => response.text()).then ((message) => {
+      document.getElementById('message-container').innerText = message;
+    });
```
Comment (aminry) [inline portfolio/src/main/webapp/script.js]: Also add error handling in case the call fails.

Comment (wait321) [reply portfolio/src/main/webapp/script.js]: Just add
.catch((error) => {
  console.error('Error:', error);
});

File: portfolio/src/main/webapp/index.html (hunk 1 above)
Comment (wait321) [reply portfolio/src/main/webapp/index.html]: Sorry, I should have been more clear. Change to `<br />`. What you changed it to is valid, but it's pointlessly long.

**Review by wait321 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/index.html (hunk 1 above)
Comment (SusanaPerezJ) [reply portfolio/src/main/webapp/index.html]: thank you, I'll make the changes

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (SusanaPerezJ) [reply portfolio/src/main/webapp/script.js]: sounds good, thanks!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -19,14 +19,61 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*;
```
Comment (wait321) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Don't import .* it makes it so difficult to figure out what your dependencies are.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -19,14 +19,61 @@
...
+import com.google.gson.Gson;
```
Comment (wait321) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: You will need a pom.xml change to download this library before  your code can find it
https://mvnrepository.com/artifact/com.google.code.gson/gson/2.8.6

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -19,14 +19,61 @@
...
+ /**
+  * Takes data from the request and populates arraylist
+  */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String text = getParameter(request, "comment-box", "");
+    //Handles empty comments, does not add them to the thread
+    if(validComment(text)){
+        comments.add(text);
+        response.setContentType("text/html");
```
Comment (wait321) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Delete this line. This shouldn't be necessary if you're sending a redirect

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -19,14 +19,61 @@
...
+ 
+  @Override 
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    response.setContentType("application/json");
+    String json = new Gson().toJson(comments);
+    response.getWriter().println(json);
+  }
+ /**
```
Comment (wait321) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: add a blank line between methods

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -19,14 +19,61 @@
...
+    return value;
+  }
+/**
+ * Validates a comment if it has at least a character that is not a white space
+ */
+  public boolean validComment(String comment){
+    if(comment.isEmpty()){
+        return false;
+      }
+    for(int j = 0; j < comment.length(); j++){
```
Comment (wait321) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: There's a library method that already does this: trim() https://www.w3schools.com/java/showjava.asp?filename=demo_ref_string_trim

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -19,14 +19,61 @@
...
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
+  }
+/**
+ * Validates a comment if it has at least a character that is not a white space
+ */
+  public boolean validComment(String comment){
+    if(comment.isEmpty()){
```
Comment (wait321) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: this entire method can be replaced with the line

return !comment.trim().isEmpty();

File: portfolio/src/main/webapp/index.html [hunk 9]
```diff
@@ -28,12 +28,22 @@ <h2>Pictures!</h2>
...
       <button onclick="addRandomFact()">Click Here!</button>
-      <div id="fact-container"></div>
+      <div id="fact-container" class="space-between-elements"></div>
       <a href="https://www.linkedin.com/in/susana-perez-juarez/">LinkedIn profile</a>
       <a href="https://github.com/SusanaPerezJ">GitHub profile</a>
+      <form action="/data" method="POST" class="space-between-elements">
+      <p>Leave a comment below!</p>
+        <textarea name="comment-box" rows="6" cols="100"></textarea>
+        <p><class="space-between-elements"></p>
+        <input type="submit">
```
Comment (wait321) [inline portfolio/src/main/webapp/index.html]: Need to close the tag. Also, this looks like a button, which is easier to understand if you use `<button type="submit" />`

File: portfolio/src/main/webapp/index.html [hunk 10]
```diff
@@ -28,12 +28,22 @@ <h2>Pictures!</h2>
...
       <p>Click this button to see a random fact about me:</p>
       <button onclick="addRandomFact()">Click Here!</button>
-      <div id="fact-container"></div>
+      <div id="fact-container" class="space-between-elements"></div>
       <a href="https://www.linkedin.com/in/susana-perez-juarez/">LinkedIn profile</a>
       <a href="https://github.com/SusanaPerezJ">GitHub profile</a>
+      <form action="/data" method="POST" class="space-between-elements">
+      <p>Leave a comment below!</p>
+        <textarea name="comment-box" rows="6" cols="100"></textarea>
+        <p><class="space-between-elements"></p>
```
Comment (wait321) [inline portfolio/src/main/webapp/index.html]: There's no HTML tag `<class ...>`. You can just add the class to textarea or the p above. Example:
`<textarea class="space-between-elements" ...`

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -14,15 +14,34 @@
...
- * Add a random photo to the page.
+ * Adds a random fact to the page.
  **/
  function addRandomFact() {
      const facts = ['I am bilingual', 'I love plants', 'I have purple hair', 'My favorite color is blue'];
      //Choose a random fact
      const randomFact = facts[Math.floor(Math.random() * facts.length)];
      
-     //Add random fact to the page
+     //Add it to the page
```
Comment (wait321) [inline portfolio/src/main/webapp/script.js]: revert this change. "random fact" is easier to understand than "it".

**Review by wait321 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (wait321) [reply portfolio/src/main/webapp/script.js]: You still need to add the catch

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (SusanaPerezJ) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: noted, thank you

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (SusanaPerezJ) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I added this dependency to the pom file already. should I commit that change too?

File: portfolio/src/main/webapp/script.js (hunk 11 above)
Comment (SusanaPerezJ) [reply portfolio/src/main/webapp/script.js]: the reason i changed it was because on the comment of the method i already mentioned that is adding a random fact so i didn't want to sound repetitive. 
Do you think it's better to revert that change or just delete the method comment altogether?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (wait321) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Yes. Please add it to the pull request

File: portfolio/src/main/webapp/script.js (hunk 11 above)
Comment (wait321) [reply portfolio/src/main/webapp/script.js]: Either is fine. I don't feel any strong preference between the 2 options.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -19,14 +19,61 @@
...
-  @Override
+  public ArrayList<String> comments = new ArrayList<>();
+ 
+  @Override 
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    response.setContentType("application/json");
+    String json = new Gson().toJson(comments);
+    response.getWriter().println(json);
```
Comment (aminry) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Have member variable for this gson instance, and don't create a new instance everytime.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -19,14 +19,61 @@
...
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/index.html");
+  }
+/**
+ * Returns the request parameter, or the default value if the parameter was not specified by the client
+ */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
```
Comment (aminry) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: use ternary operator

return value == null ? defaultValue : value;

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 12 above)
Comment (aminry) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: no need to use println(, instead simply use print(

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 14]
```diff
@@ -19,14 +19,49 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
```
Comment (aminry) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:22]: nit: Please sort the import alphabetical

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 15]
```diff
@@ -19,14 +19,49 @@
...
+import java.util.ArrayList;
+import com.google.gson.Gson;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that returns thread of comments. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  public ArrayList<String> comments = new ArrayList<>();
+  private Gson gson = new Gson();
```
Comment (aminry) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Please make this "static final"

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 16]
```diff
@@ -19,14 +19,49 @@
...
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import com.google.gson.Gson;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that returns thread of comments. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  public ArrayList<String> comments = new ArrayList<>();
```
Comment (aminry) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Please make this static final, and it can be private.

File: portfolio/src/main/webapp/index.html [hunk 17]
```diff
@@ -28,12 +28,21 @@ <h2>Pictures!</h2>
...
-      <h2>Random Facts about me </h2> 
+      <h2>Random Facts about me</h2> 
       <p>Click this button to see a random fact about me:</p>
       <button onclick="addRandomFact()">Click Here!</button>
-      <div id="fact-container"></div>
+      <div id="fact-container" class="space-between-elements"></div>
       <a href="https://www.linkedin.com/in/susana-perez-juarez/">LinkedIn profile</a>
       <a href="https://github.com/SusanaPerezJ">GitHub profile</a>
+      <form action="/data" method="POST" class="space-between-elements">
+      <p>Leave a comment below!</p>
```
Comment (aminry) [inline portfolio/src/main/webapp/index.html:37]: nit: you want tap this forward to be consistent

File: portfolio/src/main/webapp/index.html [hunk 18]
```diff
@@ -28,12 +28,21 @@ <h2>Pictures!</h2>
...
+        <textarea name="comment-box" rows="6" cols="100"></textarea>
+        <button type="submit">Submit comment</button>
+        <p>Comments:</p>
```
Comment (aminry) [inline portfolio/src/main/webapp/index.html]: This portion does not need to be inside the form tag.

**Review by aminry — APPROVED** [review]

File: portfolio/src/main/webapp/index.html (hunk 17 above)
Comment (wait321) [reply portfolio/src/main/webapp/index.html:37]: All lines between an open tag like `<form>` and its closing tag `</form>` should have extra spaces in front (like what you did for the line after `<body>`)

**Review by wait321 — APPROVED** [review]
